python/lsst/ts/ess/common/data_client/base_read_loop_data_client.py

```diff
@@ -109,67 +175,65 @@ async def disconnect(self) -> None:
-        except asyncio.InvalidStateError:
-            self.log.warning("Task is still running so not raising the exception.")
+        """Perform the life cycle of a DtaClient in a loop.
 
-    async def read_loop(self) -> None:
-        """Call the read_data method in a loop.
+        The life cycle is:
 
-        Allow for max_timeouts timeouts before raising a TimeoutError.
+          - connect
+          - setup reading
+          - read data while connected
+          - handle exceptions
+          - disconnect
         """
-        await self.setup_reading()
-        # Number of consecutive read timeouts encountered.
         self.num_consecutive_read_timeouts = 0
-        while self.connected:
-            rate_limit_task = asyncio.create_task(asyncio.sleep(self.rate_limit))
+        self.timeout_event.clear()
+        while True:
             try:
-                await self.read_data_once()
-            finally:
-                await rate_limit_task
+                if not self.connected:
+                    await self.connect()
+                await self.setup_reading()
+                while self.connected:
+                    rate_limit_task = asyncio.create_task(
+                        asyncio.sleep(self.rate_limit)
+                    )
+                    try:
+                        await self.read_data()
+                        self.num_consecutive_read_timeouts = 0
+                        self.timeout_event.clear()
+                    finally:
+                        await rate_limit_task
+            except asyncio.CancelledError:
+                self.loop_should_end = True
+            except Exception as e:
+                self.num_consecutive_read_timeouts += 1
+                if self.num_consecutive_read_timeouts >= self.max_read_timeouts:
+                    self.log.error(
+                        f"Read timed out {self.num_consecutive_read_timeouts} times "
+                        f">= {self.max_read_timeouts=}; giving up. Raising {e!r}."
+                    )
+                    self.loop_should_end = True
+                    self.timeout_event.set()
+                    raise
 
-    async def read_data_once(self) -> None:
-        try:
-            await self.read_data()
-            self.num_consecutive_read_timeouts = 0
-        except asyncio.TimeoutError:
-            self.num_consecutive_read_timeouts += 1
-
-            if self.num_consecutive_read_timeouts >= self.config.max_read_timeouts:
-                self.log.error(
-                    f"Read timed out {self.num_consecutive_read_timeouts} times "
-                    f">= {self.config.max_read_timeouts=}; giving up."
+                message = (
+                    f"Read timed out. This is timeout #{self.num_consecutive_read_timeouts} "
+                    f"of {self.max_read_timeouts} allowed. Error was: {e!r}."
+                    " Attempting to disconnect and reconnect now."
```

Comment:
Please change that to "Attempting to reconnect." reconnect implies disconnection. Also makes the message more dense and information rich. Thanks.